retrieval/discovery/gce.go

```diff
@@ -0,0 +1,171 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package discovery
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"google.golang.org/api/compute/v1"
+
+	"github.com/prometheus/common/log"
+	"github.com/prometheus/common/model"
+	"golang.org/x/net/context"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/google"
+
+	"github.com/prometheus/prometheus/config"
+)
+
+const (
+	gceLabel             = model.MetaLabelPrefix + "gce_"
+	gceLabelProject      = gceLabel + "project"
+	gceLabelZone         = gceLabel + "zone"
+	gceLabelNetwork      = gceLabel + "network"
+	gceLabelSubnetwork   = gceLabel + "subnetwork"
+	gceLabelPublicIP     = gceLabel + "public_ip"
+	gceLabelPrivateIP    = gceLabel + "private_ip"
+	gceLabelInstanceName = gceLabel + "instance_name"
+	gceLabelTags         = gceLabel + "tags"
+)
+
+// GCEDiscovery periodically performs GCE-SD requests. It implements
+// the TargetProvider interface.
+type GCEDiscovery struct {
+	project      string
+	zone         string
+	filter       string
+	client       *http.Client
+	svc          *compute.Service
+	isvc         *compute.InstancesService
+	interval     time.Duration
+	port         int
+	tagSeparator string
+}
+
+// NewGCEDiscovery returns a new GCEDiscovery which periodically refreshes its targets.
+func NewGCEDiscovery(conf *config.GCESDConfig) *GCEDiscovery {
+	gd := &GCEDiscovery{
+		project:      conf.Project,
+		zone:         conf.Zone,
+		filter:       conf.Filter,
+		interval:     time.Duration(conf.RefreshInterval),
+		port:         conf.Port,
+		tagSeparator: conf.TagSeparator,
+	}
+	var err error
+	gd.client, err = google.DefaultClient(oauth2.NoContext, compute.ComputeReadonlyScope)
+	if err != nil {
+		log.Errorf("error setting up communication with GCE service: %s", err)
+		return gd
+	}
+	gd.svc, err = compute.New(gd.client)
+	if err != nil {
+		log.Errorf("error setting up communication with GCE service: %s", err)
+		return gd
+	}
+	gd.isvc = compute.NewInstancesService(gd.svc)
+	return gd
+}
+
+// Run implements the TargetProvider interface.
+func (gd *GCEDiscovery) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
+	defer close(ch)
+
+	ticker := time.NewTicker(gd.interval)
+	defer ticker.Stop()
+
+	// Make sure gce setup was done correctly
+	if gd.isvc == nil {
+		log.Error(errors.New("GCE communication setup failed"))
+		return
+	}
+
+	// Get an initial set right away.
+	tg, err := gd.refresh()
+	if err != nil {
+		log.Error(err)
+	} else {
+		ch <- []*config.TargetGroup{tg}
+	}
+
+	for {
+		select {
+		case <-ticker.C:
+			tg, err := gd.refresh()
+			if err != nil {
+				log.Error(err)
+			} else {
+				ch <- []*config.TargetGroup{tg}
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (gd *GCEDiscovery) refresh() (*config.TargetGroup, error) {
+	if gd.isvc == nil {
+		return nil, errors.New("could not refresh gce targets because of previous gce communication setup error")
```

Comment:
Or you return an error in the constructor and don't register the discovery at all if the setup fails. It would never be valid to use a discovery type without `isvc` set then. The kubernetes discovery does something similar.